.github/workflows/ci.yml

```diff
@@ -0,0 +1,58 @@
+name: CI
+
+on:
+  push:
+    # Avoid duplicate builds on PRs.
+    branches:
+      - main
+    tags: 
+      - v*
+  pull_request:
+
+permissions:
+  contents: read
+
+jobs:
+  shellcheck:
+    runs-on: ubuntu-22.04
+    container:
+      image: koalaman/shellcheck-alpine:v0.8.0
+    steps:
+    - uses: actions/checkout@v2
+    - name: Run shellcheck
+      run: find . -type f \( -name "*.sh" -o -path "*/bin/*" \) ! -name '*.jq' | xargs -t shellcheck
+  build-stack:
+    runs-on: ubuntu-22.04
+    needs:
+    - shellcheck
+    env:
+      STACK: heroku-${{ matrix.stack-version }}
+      STACK_VERSION: "${{ matrix.stack-version }}"
+      DOCKER_HUB_TOKEN: "${{ secrets.DOCKER_HUB_TOKEN }}"
+      ID_SERVICE_PASSWORD: "${{ secrets.ID_SERVICE_PASSWORD }}"
+      MANIFEST_APP_TOKEN: "${{ secrets.MANIFEST_APP_TOKEN }}"
+    strategy:
+      matrix:
+        stack-version:
+        - '18'
+        - '20'
+        - '22'
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build images
+      run: bin/build.sh $STACK_VERSION
+    - name: Check that the generated files are in sync
+      run: |-
+        status="$(git status --porcelain)"
+        if [[ -n "$status" ]]; then
+          echo "Generated files differ from checked-in versions! Run bin/build.sh to regenerate them locally."
+          echo -e "\nChanged files:\n${status}\n"
+          git diff
+          exit 1
+        fi
+    - name: Publish to image registries
+      run: bin/publish-to-registries.sh
+      if: "'main' == ${{ github.ref }} || ${{ github.ref }}"
```

Comment:
The publishing step is occurring on this PR branch at the moment, rather than just main (it didn't succeed, since I intentionally left the creds unset in the repo settings, just in case). 

The builder repo does this for only running on main:
https://github.com/heroku/builder/blob/01659d4b7bc72eb923c45a3c08033b27dad9a42a/.github/workflows/build-test-publish.yml#L122

Though this repo also needs this step to run on tags too, so it will need tweaking :-)

Reply:
I used the github.ref_type to check the ref is a tag, and the github.ref_name to assert it's main